### PR TITLE
forge(verify): sourcify support

### LINKED ISSUE #1822: Sourcify support
### Component

Forge

### Describe the feature you would like

I have checked a several ways to verify codes for [blockscout](https://github.com/blockscout/blockscout/releases/tag/v4.1.3-beta) v4.1.3 or above. There is no api-key needed for [verification](https://docs.blockscout.com/for-users/verifying-a-smart-contract/contracts-verification-via-sourcify). Currently there is a way to upload the standard json for the process however the json that created from the regular `forge` does not work. 

It should be nice to have blockscout specific solution for source verifications with this flag `--blockscout`

for example:
```
forge create --rpc-url "https://rpc1.oasischain.io" \
  --private-key "...." vault/stakingpoolv3/Woac.sol:Woac \
  --verify --legacy --blockscout

```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by rkrasiuk**
close #1822 

kudos to @robertabbott for kicking this off

### CODE REVIEW

File: cli/src/cmd/forge/verify/etherscan.rs [hunk 1]
```diff
@@ -34,108 +26,26 @@ use std::{
...
-        help = "The project's root path. By default, this is the root directory of the current Git repository or the current working directory if it is not part of a Git repository",
-        long,
-        value_hint = ValueHint::DirPath,
-        value_name = "PATH"
-    )]
-    pub root: Option<PathBuf>,
-}
-
-impl_figment_convert_basic!(VerifyArgs);
+pub struct EtherscanVerificationProvider {}
```
Comment (gakonst) [inline cli/src/cmd/forge/verify/etherscan.rs:36]: nit 

```suggestion
pub struct EtherscanVerificationProvider;
```

File: cli/src/cmd/forge/verify/mod.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+pub enum VerificationProviderType {
+    Etherscan,
+    Sourcify,
+}
+
+impl VerificationProviderType {
+    fn client(&self) -> Box<dyn VerificationProvider> {
+        match self {
+            VerificationProviderType::Etherscan => Box::new(EtherscanVerificationProvider {}),
+            VerificationProviderType::Sourcify => Box::new(SourcifyVerificationProvider {}),
```
Comment (gakonst) [inline cli/src/cmd/forge/verify/mod.rs:179]:
```suggestion
            VerificationProviderType::Etherscan => Box::new(EtherscanVerificationProvider),
            VerificationProviderType::Sourcify => Box::new(SourcifyVerificationProvider),
```

File: cli/src/cmd/forge/verify/sourcify.rs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        let path = match args.contract.path {
+            Some(path) => PathBuf::from(path),
+            None if config.cache => {
+                let cache = project.read_cache_file()?;
+                crate::cmd::get_cached_entry_by_name(&cache, &args.contract.name)?.0
+            }
+            _ => eyre::bail!(
+                "If cache is disabled, contract info must be provided in the format <path>:<name>"
+            ),
+        };
```
Comment (gakonst) [inline cli/src/cmd/forge/verify/sourcify.rs:56]: for separate PR but this "read artifact from name" logic should go behind a helper fn

**Review by gakonst — APPROVED** [review]
couple nits and needs linting, but nice work overall. cc @robertabbott FYI this is now over the line if you wanna take a look at the impl

**Comment by selarun13** [comment]
0x6F99E90Dd385c23826b416528e325214671FeBf6

**Pull request merged**